apiserver/authentication/jwt/jwt.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package jwt
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+
+	"github.com/juju/juju/apiserver/authentication"
+	"github.com/juju/juju/core/permission"
+)
+
+// JWTTokenEntity represents the entity found within a JWT token and conforms to
+// state.Entity
+type JWTTokenEntity struct {
+	user names.UserTag
+}
+
+// Tag implements state.Entity
+func (t JWTTokenEntity) Tag() names.Tag {
+	return t.user
+}
+
+type Authenticator interface {
+	authentication.RequestAuthenticator
+	JWTTokenParser
+}
+
+// TokenParser parses a jwt token returning the token and
+// entity derived from the token subject.
+type JWTTokenParser interface {
+	Parse(ctx context.Context, tok string) (jwt.Token, authentication.Entity, error)
+}
+
+// JWTAuthenticator is an authenticator responsible for handling JWT tokens from
+// a client.
+type JWTAuthenticator struct {
+	cache      *jwk.Cache
+	httpClient *http.Client
+	refreshURL string
+}
+
+const (
+	wellKnownPath = "/.well-known/jwks.json"
+)
+
+func NewJWTAuthenticator(refreshURL string) *JWTAuthenticator {
+	httpClient := &http.Client{
+		Timeout: 30 * time.Second,
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+
+	return NewJWTAuthenticatorWithHTTPClient(httpClient, refreshURL)
+}
+
+func NewJWTAuthenticatorWithHTTPClient(
```

Comment:
This is only called here, do we need it? Or is it for tests?

Reply:
Helps with tests.